src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java

```diff
@@ -200,58 +216,59 @@ private MethodHandle specialize(MethodHandle leafHandle) {
 
         MethodHandle specializedHandle = leafHandle; // initial
 
-        int insertPos = -1;
+        int argInsertPos = -1;
```

Comment:
The more we tweak this routine, the more I'm convinced we should just spin specialized bytecode :-) - it's becoming increasingly hard to follow what's happening here. (not for this review - more longer term)

Reply:
Yeah, especially the indexing is tricky since we have to walk the recipe in reverse, but we shouldn't have that problem when spinning bytecode.